### PR TITLE
[FLINK-14516][network] Remove non credit-based flow control code

### PR DESCRIPTION
**Pull request by pnowojski**
This PR removes non credit-based flow control code

## Verifying this change

*(Please pick either of the following options)*

This change is covered by existing tests. Some of them needed to be adjusted after removing no longer used classes.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerAlignmentLimitTest.java [hunk 1]
```diff
@@ -354,4 +363,83 @@ public void describeTo(Description description) {
...
+
+		public List<Long> getAndResetTriggeredCheckpoints() {
+			List<Long> copy = new ArrayList<>(triggeredCheckpoints);
+			triggeredCheckpoints.clear();
+			return copy;
+		}
+	}
+
+	private static class AssertCheckpointWasAborted implements Closeable {
+		private CheckpointNotificationVerifier checkpointNotificationVerifier;
```
Comment (zhijiangW) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerAlignmentLimitTest.java]: nit: final

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerAlignmentLimitTest.java [hunk 2]
```diff
@@ -251,8 +260,9 @@ public void testAlignmentLimitWithQueuedAlignments() throws Exception {
 		// checkpoint 4 completed - check and validate buffered replay
 		check(sequence[9], buffer.pollNext().get());
 		validateAlignmentTime(startTs, buffer);
-		verify(toNotify, times(1)).triggerCheckpointOnBarrier(
-			argThat(new CheckpointMatcher(4L)), any(CheckpointOptions.class), any(CheckpointMetrics.class));
```
Comment (zhijiangW) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerAlignmentLimitTest.java]: The `CheckpointMatcher` class can also be removed as a result.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironmentTest.java [hunk 3]
```diff
@@ -184,17 +158,17 @@ private void testRegisterTaskWithLimitedBuffers(int bufferPoolSize) throws Excep
...
 
 		assertEquals(Integer.MAX_VALUE, ig1.getBufferPool().getMaxNumberOfMemorySegments());
 		assertEquals(Integer.MAX_VALUE, ig2.getBufferPool().getMaxNumberOfMemorySegments());
-		assertEquals(enableCreditBasedFlowControl ? floatingBuffers : expectedBuffers, ig3.getBufferPool().getMaxNumberOfMemorySegments());
-		assertEquals(enableCreditBasedFlowControl ? floatingBuffers : expectedRp4Buffers, ig4.getBufferPool().getMaxNumberOfMemorySegments());
+		assertEquals(floatingBuffers, ig3.getBufferPool().getMaxNumberOfMemorySegments());
+		assertEquals(floatingBuffers, ig4.getBufferPool().getMaxNumberOfMemorySegments());
 
-		int invokations = enableCreditBasedFlowControl ? 1 : 0;
+		int invokations = true ? 1 : 0;
```
Comment (zhijiangW) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironmentTest.java]: nit: `int invokations = 1` directly

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerMassiveRandomTest.java [hunk 4]
```diff
@@ -60,7 +55,12 @@ public void testWithTwoChannelsAndRandomBarriers() {
 					new BufferPool[] { pool1, pool2 },
 					new BarrierGenerator[] { new CountBarrier(100000), new RandomBarrier(100000) });
 
-			CheckpointedInputGate checkpointedInputGate = new CheckpointedInputGate(myIG, new BufferSpiller(ioMan, PAGE_SIZE), "Testing: No task associated", null);
+			CheckpointedInputGate checkpointedInputGate =
+				new CheckpointedInputGate(
+					myIG,
+					new CachedBufferStorage(PAGE_SIZE, -1, "Testing"),
```
Comment (zhijiangW) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierAlignerMassiveRandomTest.java]: nit: `new CachedBufferStorage(PAGE_SIZE)` instead

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 5]
```diff
@@ -35,11 +40,31 @@
...
 @Internal
-public class CachedBufferStorage extends AbstractBufferStorage {
+public class CachedBufferStorage implements BufferStorage {
+	private static final Logger LOG = LoggerFactory.getLogger(CachedBufferStorage.class);
+
+	/**
+	 * The pending blocked buffer/event sequences. Must be consumed before requesting further data
+	 * from the input gate.
+	 */
+	protected final ArrayDeque<BufferOrEventSequence> queuedBuffered = new ArrayDeque<>();
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: can be private, also for the other following variables.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 6]
```diff
@@ -35,11 +40,31 @@
...
 	/** The page size, to estimate the total cached data size. */
 	private final int pageSize;
 
+	/**
+	 * The sequence of buffers/events that has been unblocked and must now be consumed before
+	 * requesting further data from the input gate.
+	 */
+	protected BufferOrEventSequence currentBuffered;
+
+	/** The number of bytes in the queued spilled sequences. */
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: spilled->cached

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 7]
```diff
@@ -74,49 +102,139 @@ public void add(BufferOrEvent boe) {
...
 
 	/**
-	 * It is never reusing resources and is defaulting to {@link #rollOverWithoutReusingResources()}.
+	 * Starts a new sequence of buffers and event reusing the same resources and
+	 * returns the current sequence of buffers for reading.
+	 *
+	 * @return The readable sequence of buffers and events, or 'null', if nothing was added.
 	 */
-	@Override
 	public BufferOrEventSequence rollOverReusingResources() {
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: It is no need to distinguish `rollOverReusingResources` and `rollOverWithoutReusingResources` if we only have `CachedBufferStorage`?  If so we can retain only one method, also for private method instead.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 8]
```diff
@@ -129,21 +247,30 @@ public long getPendingBytes() {
...
+		BufferOrEventSequence(ArrayDeque<BufferOrEvent> buffers, long size) {
 			this.queuedBuffers = buffers;
 			this.size = size;
 		}
 
-		@Override
+		/**
+		 * Initializes the sequence for reading.
+		 */
 		public void open() {}
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: nit: can be removed.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 9]
```diff
@@ -74,49 +102,139 @@ public void add(BufferOrEvent boe) {
...
+		}
+		// TODO: FLINK-12536 for non credit-based flow control, getNext method is blocking
+		Optional<BufferOrEvent> next = Optional.ofNullable(currentBuffered.getNext());
+		if (!next.isPresent()) {
+			completeBufferedSequence();
+		}
+		return next;
+	}
+
+	protected void completeBufferedSequence() {
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: can be private

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java [hunk 10]
```diff
@@ -74,49 +102,139 @@ public void add(BufferOrEvent boe) {
...
+	public boolean isEmpty() {
+		return currentBuffered == null;
+	}
+
+	@Override
+	public Optional<BufferOrEvent> pollNext() {
+		if (currentBuffered == null) {
+			return Optional.empty();
+		}
+		// TODO: FLINK-12536 for non credit-based flow control, getNext method is blocking
```
Comment (zhijiangW) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CachedBufferStorage.java]: This `TODO` is invalid now and can be removed?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java [hunk 11]
```diff
@@ -180,18 +176,14 @@ private static void registerInputMetrics(
...
-			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
-			CreditBasedInputBuffersUsageGauge creditBasedInputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
-				floatingBuffersUsageGauge,
-				exclusiveBuffersUsageGauge,
-				inputGates);
-			buffersGroup.gauge(METRIC_INPUT_EXCLUSIVE_BUFFERS_USAGE, exclusiveBuffersUsageGauge);
-			buffersGroup.gauge(METRIC_INPUT_FLOATING_BUFFERS_USAGE, floatingBuffersUsageGauge);
-			buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, creditBasedInputBuffersUsageGauge);
-		} else {
-			buffersGroup.gauge(METRIC_INPUT_POOL_USAGE, new InputBufferPoolUsageGauge(inputGates));
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java]: also remove `InputBufferPoolUsageGauge` class

**Comment by pnowojski** [comment]
Thanks @zhijiangW for the review. I've updated the code and added one more cleaning up commit for some renames.

**Comment by zhijiangW** [comment]
Yes I had the same feeling with you for `currentBuffers` and `currentBuffered` while reviewing. Considering it is out of this PR scope, so I did not mention it before. We could refactor it further if have time future.:)

**Comment by pnowojski** [comment]
Thanks for the review @zhijiangW. Merging

**Pull request merged**